Web Inspector: Broken Sources and Network tabs after creating Local Override with invalid Regular Expression

rcaliman-apple opened the pull request:
#### 1d01240dc0835e67852aa75ffae66327316e9395
<pre>
Web Inspector: Broken Sources and Network tabs after creating Local Override with invalid Regular Expression
<a href="https://bugs.webkit.org/show_bug.cgi?id=294072">https://bugs.webkit.org/show_bug.cgi?id=294072</a>
<a href="https://rdar.apple.com/152651800">rdar://152651800</a>

Reviewed by Devin Rousso.

Prevent creating a Local Override with an invalid URL regular expression
to avoid triggering an exception when it is applied.

Since Local Overrides are stored and re-applied on every Web Inspector
session, this unhandled exception triggered on every URL request
makes the Sources and Network tabs unusable for every inspected page.

With the Sources tab in a broken state, a user cannot even remove
the faulty Local Override.

For users who end up in this state, we include a mitigation to automatically
delete Local Overrides with broken regular expressions as URLs.

* Source/WebInspectorUI/UserInterface/Models/LocalResourceOverride.js:
(WI.LocalResourceOverride.prototype.matches):
(WI.LocalResourceOverride.prototype.get _urlRegex):
(WI.LocalResourceOverride):

Canonical link: <a href="https://commits.webkit.org/295928@main">https://commits.webkit.org/295928@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ae28dcaed09fd9e79922009d8c3efaa21df12ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/106597 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26348 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16745 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111805 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/57193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27018 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34850 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/80952 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/57193 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/109601 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/21432 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/96190 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61288 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/20884 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/14296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/56638 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90771 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/14326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/114666 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33735 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24873 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/90021 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34099 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/92420 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89730 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22897 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34640 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/12453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/29362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33660 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/39073 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33406 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36759 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35005 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/LocalResourceOverride.js:
```diff
@@ -346,7 +346,11 @@ WI.LocalResourceOverride = class LocalResourceOverride extends WI.Object
         console.assert(this._isRegex);
-        return new RegExp(this._url, !this._isCaseSensitive ? "i" : "");
+        try {
+            return new RegExp(this._url, !this._isCaseSensitive ? "i" : "");
```
this seems like the wrong place to check.  we shouldn't even have a `WI.LocalResourceOverride` with invalid information in the first place (i.e. would it just never match anything and effectively be silently ignored?)

we already do other validation inside `WI.LocalResourceOverridePopover.prototype.get serializedData` so i think we should have something like this in there
```js
if (data.isRegex) {
    try {
        new RegExp(data.url);
    } catch {
        return null;
    }
}
```

rcaliman-apple replied:
>(i.e. would it just never match anything and effectively be silently ignored?)

Yes, that's what happens.

The user experience isn't great as-is, but it allows a developer to correct the mistake.

If we return `null` for `WI.LocalResourceOverridePopover.prototype.serializedData`, the Local Override won't even be created in `WI.LocalResourceOverrideTreeElement.willDismissPopover()` without any warning why.

One extra reason for keeping the fix here is to recover Web Inspector out of a bad state for people who ended up committing the bad RegEx to storage, like the [original reporter](https://developer.apple.com/forums/thread/772801).

rcaliman-apple replied:
If anything, I'd keep the solution in both places. 

Having an unguarded `new RegExp()` in `WI.LocalResourceOverride` is bound to bite again in the future if we offer another way to create it instead of free-form user input.

dcrousso replied:
the typical pattern in Web Inspector is that by the time we're in the `constructor` (and ideally even earlier than that like in the "view" layer) all of the arguments should already have been checked to be valid

if the developer has something in their storage that's invalid, id suggest we check it there

as such, i think we instead should do something like this (in addition to the change you already made in `WI.LocalResourceOverridePopover.prototype.get serializedData`)
```js
let localResourceOverride = WI.LocalResourceOverride.fromJSON(serializedLocalResourceOverride);

if (localResourceOverride.isRegex) {
    // Temporary fix for if the local override was created before invalid URL regex checking was added.
    try {
        localResourceOverride._urlRegex;
    } catch {
        const key = null;
        WI.objectStores.localResourceOverrides.associateObject(localResourceOverride, key, serializedLocalResourceOverride);
        WI.objectStores.localResourceOverrides.deleteObject(localResourceOverride);
    }
}
```

rcaliman-apple replied:
This makes sense and helps to restore Web Inspector for users who accidentally ended up in this state. Updated PR 👍

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/NetworkManager.js:
```diff
@@ -73,6 +73,18 @@ WI.NetworkManager = class NetworkManager extends WI.Object
                     let localResourceOverride = WI.LocalResourceOverride.fromJSON(serializedLocalResourceOverride);
 
+                    if (localResourceOverride.isRegex) {
+                        // Fix for stored local overrides created before URL regex checking was added. See <https://webkit.org/b/294072>
```
we probably should have this be a new bug for eventually removing this as we ideally won't need this forever
```
// FIXME <https://webkit.org/b/######> Remove fix for stored local overrides created before URL regex checking was added
```
and you can have that bug be blocked by <https://webkit.org/b/294072>

rcaliman-apple replied:
Done! Filed https://bugs.webkit.org/show_bug.cgi?id=294126

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/LocalResourceOverridePopover.js:
```diff
@@ -77,6 +77,12 @@ WI.LocalResourceOverridePopover = class LocalResourceOverridePopover extends WI.
             const schemes = ["http:", "https:", "file:"];
             if (!schemes.some((scheme) => data.url.toLowerCase().startsWith(scheme)))
                 return null;
+        } else {
```
NIT: it'd probably flip the order of this `if` so that the `isRegex` check is closer to the `RegExp` usage
```js
if (data.isRegex) {
    try {
        new RegExp(data.url);
    } catch {
        return null;
    }
} else {
    const schemes = ["http:", "https:", "file:"];
    if (!schemes.some((scheme) => data.url.toLowerCase().startsWith(scheme)))
        return null;
}
```
apologies for not suggesting this before.  didn't realize we already had a check for `data.isRegex` nearby

rcaliman-apple replied:
Funny enough, I considered changing it like this the first time around, but I hesitated 😄 
Updated now.

rcaliman-apple committed:
Web Inspector: Broken Sources and Network tabs after creating Local Override with invalid Regular Expression
https://bugs.webkit.org/show_bug.cgi?id=294072
rdar://152651800

Reviewed by Devin Rousso.

Prevent creating a Local Override with an invalid URL regular expression
to avoid triggering an exception when it is applied.

Since Local Overrides are stored and re-applied on every Web Inspector
session, this unhandled exception triggered on every URL request
makes the Sources and Network tabs unusable for every inspected page.

With the Sources tab in a broken state, a user cannot even remove
the faulty Local Override.

For users who end up in this state, we include a mitigation to automatically
delete Local Overrides with broken regular expressions as URLs.

* Source/WebInspectorUI/UserInterface/Models/LocalResourceOverride.js:
(WI.LocalResourceOverride.prototype.matches):
(WI.LocalResourceOverride.prototype.get _urlRegex):
(WI.LocalResourceOverride):

Canonical link: https://commits.webkit.org/295928@main
